File: src/com/turlington/JobChecker.java
package com.turlington;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;

import java.util.Random;
import java.util.Set;


/**
 * Checks for job postings.
 * Created by Mitchell on 7/18/2016.
 */
public class JobChecker implements iJobChecker {
	private final Set<JobSite> jobSites;
    private final Set<String> jobListings;
    private final FileLoader fileLoader;
    private final iNotify notifier;

    JobChecker(Set<JobSite> jobSites, FileLoader fileLoader, iNotify notifier) {
        this.notifier = notifier;
        this.fileLoader = fileLoader;
        this.jobListings = fileLoader.getJobListingsFromFile();
        this.jobSites = jobSites;
		WebDriver driver = new FirefoxDriver();
        for (JobSite jobSite : jobSites) {
            jobSite.setWebDriver(driver);
        }
    }

    @Override
    public void checkSites() {
    	Random random = new Random();
        //noinspection InfiniteLoopStatement
        while (true) { //TODO: maybe should give a real way to exit the loop aside from terminating program manually.
            for (JobSite jobSite : jobSites) {
                jobSite.goToPage();
                checkJobs(jobSite);
            }
            double minSeconds = random.nextDouble() + random.nextInt(3);
            Main.waitMillis((long) (minSeconds * 60 * 1000));
        }
    }

    @Override
    public void checkJobs(JobSite jobsite) {
        Set<JobListing> jobsFound = jobsite.getJobListings();
        jobsFound.stream().filter(jobListing -> !jobListings.contains(jobListing.getTitle())).forEach(jobListing -> {
            jobListings.add(jobListing.getTitle());
            fileLoader.saveFile(jobListing);
            notifier.announce(jobListing);
        });
    }
}

File: src/main/kotlin/com/turlington/JobChecker.kt
package com.turlington

import org.apache.log4j.Logger
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*


/**
 * Checks for job postings.
 * Created by Mitchell on 7/18/2016.
 */
class JobChecker internal constructor(private val jobSites: Set<JobSite>, private val fileLoader: FileLoader, private val notifier: iNotify) : iJobChecker {
    private val logger = Logger.getLogger(javaClass)
    private val jobListings: MutableSet<String> = fileLoader.jobListingsFromFile as MutableSet<String>

    override fun checkSites() {
        val random = Random()
        val stopPath = Paths.get("stop.stop")
        while (Files.notExists(stopPath)) {
            try {
                for (jobSite in jobSites) {
                    jobSite.goToPage()
                    checkJobs(jobSite)
                }
            } catch (e: Exception) {
                logger.error("Error checking sites: ${e.message}", e)
            }
            val minSeconds = random.nextDouble() + random.nextInt(3)
            Main.waitMillis((minSeconds * 60.0 * 1000.0).toLong())
        }
    }

    override fun checkJobs(jobSite: JobSite) {
        try {
            val jobsFound = jobSite.getJobListings()
            jobsFound.filter { !jobListings.contains(it.title) }.forEach { jobListing ->
                jobListings.add(jobListing.title)
                fileLoader.saveFile(jobListing)
                notifier.announce(jobListing)
            }
        } catch (e: Exception) {
            logger.error("Error getting jobs: ${e.message}", e)
        }
    }
}
